lisa/tests/base.py

```diff
@@ -905,10 +931,33 @@ def rtapp_profile(self):
     @property
     def rtapp_tasks(self):
         """
-        Sorted list of rtapp task names, as defined in ``rtapp_profile``
-        attribute.
+        The rtapp task names as found from the trace in this bundle.
+
+        :return: the list of actual trace task names
         """
-        return sorted(self.rtapp_profile.keys())
+        task_map = getattr(self, 'task_map', None)
```

Comment:
This kind of warts also go away if we just put the code that computes the task directly. Extra state means more state that can be broken. Here, `self.rtapp_tasks` would fail if called before `get_trace()`, which is broken.

Reply:
Sure, but these functions are not necessarily the first one we use.
Once we have a RTATestBundle we can either start by looking at rtapp tasks or doing some trace analysis.

Perhaps I should just keep the `task_map` generation in a memoized function, as it was at the beginning, and use it here to get rid of this check. As well as calling it from `get_trace()`.

Reply:
Moreover, failing with an exception does not mean the code is "broken".

At the very least not more broken then failing with:
```
AttributeError: 'RTABundleTest' object has no attribute 'task_map'
```
that we get if we inline the `task_map` creation only here.

Reply:
> Moreover, failing with an exception does not mean the code is "broken".

In this case the API is broken as it requires for client code to jump through hoops without any good reason. Accessing `self.rtapp_tasks` is not supposed to fail if it can succeed, regardless of `get_trace()` being called before or not.

Also, it's genuinely broken as parameters passed to `get_trace()` are not controlled, which means someone would be free to pass `window` and therefore make the task detection fail if the window is not large enough.